### PR TITLE
Add the AMKO docs

### PR DESCRIPTION
**Pull request by hemantavi**
- Main doc in docs/AMKO.
- Follow up docs in:
* docs/AMKO/crds/
* docs/AMKO/kubeconfig.md

### CODE REVIEW

File: docs/AMKO/README.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+# AMKO: Avi Multicluster Kubernetes Operator
+
+### Run AMKO
+
+AMKO is a kubernetes operator which orchestrates application deployments across multiple kubernetes/openshift clusters.
```
Comment (sudswasavi) [inline docs/AMKO/README.md:5]: AMKO is a kubernetes operator does multi-cluster application load balancing for Kubernetes workloads.

File: docs/AMKO/README.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+ ![Alt text](images/amko_ss.png?raw=true "amko architecture")
+
+AMKO is aware of the following object types:
```
Comment (sudswasavi) [inline docs/AMKO/README.md:8]: AMKO listens to the following objects:
Kubernetes:
 - Ingress
 - Service of type LoadBalancer

Openshift:
 - Routes
 - Service of type LoadBalancer

File: docs/AMKO/README.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+* Kubernetes Ingresses
+* Openshift Routes
+* Service type Load Balancer
+
+Based on a couple of CRDs `GSLBConfig` and `GlobalDeploymentPolicy`, AMKO manages the GSLB service configuration on a AVI leader site.
```
Comment (sudswasavi) [inline docs/AMKO/README.md:13]: No need of these lines.

File: docs/AMKO/README.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+ 
+#### Pre-requisites
+To kick-start AMKO, we need:
+1. Atleast one kubernetes/openshift cluster.
+2. Atleast one AVI controller which manages the kubernetes/openshift cluster(s) designated as the GSLB leader site. Additional controllers managing openshift/kubernetes clusters can be added as follower sites.
```
Comment (sudswasavi) [inline docs/AMKO/README.md:18]: s/which/that
AMKO shouldn't need OpenShift/Kubernetes virtual services to be managed in Avi.

File: docs/AMKO/README.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+3. AMKO assumes that it has connectivity to all the member clusters' kubernetes API servers. Without this, AMKO won't be able to watch over the ingress/route/services in the member kubernetes clusters.
```
Comment (sudswasavi) [inline docs/AMKO/README.md:19]: Can we separate out routes/ingress more distinctly through out the document? Let's assume that a kubernetes user may not be aware of how openshift behaves.

File: docs/AMKO/README.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+$ kubectl get gc -n avi-system gc-1
+NAME            AGE
+gc-1            45m
+
+$ kubectl get gdp -n avi-system
+NAME         AGE
+global-gdp   46m
+```
+
+*Note* that, only one instance of each type is supported and AMKO *will* ignore other objects.
```
Comment (sudswasavi) [inline docs/AMKO/README.md:120]: And reflect the status to say that a GDP/GC is rejected?

File: docs/AMKO/README.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+#### Editing runtime parameters of AMKO
+The `GDP` object can be edited in the runtime to change the application selection parameters, traffic split and the applicable clusters. AMKO will recognize these changes and will update the GSLBServices accordingly.
```
Comment (sudswasavi) [inline docs/AMKO/README.md:123]: s/in the runtime/at runtime

File: docs/AMKO/README.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+Not all fields in `GSLBConfig` can be changed in the runtime. The only allowed field is `logLevel`. This means that if the `logLevel` changes while AMKO is running, the changes will take effect. For any changes to the other fields, the AMKO pod has to be restarted to recognize the changes.
```
Comment (sudswasavi) [inline docs/AMKO/README.md:125]: s/in the runtime/at runtime

Comment (sudswasavi) [reply docs/AMKO/README.md:125]: Changing `logLevel` is only permissible at runtime via an edit of the `GSLBConfig`. For all other changes, the AMKO to the GSLBConfig, the AMKO pod has to be restarted.

File: docs/AMKO/crds/gdp.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
+## GlobalDeploymentPolicy CRD for AMKO
+A CRD called GlobalDeploymentPolicy allows users to select kubernetes/openshift objects based on certain rules. This GDP object has to be created on the same system wherever the GSLBConfig object was created and `amko` is running. The selection policy applies to all the clusters which are mentioned in the GDP object. A typical GlobalDeploymentPolicy looks like this:
```
Comment (sudswasavi) [inline docs/AMKO/crds/gdp.md:2]: You have introduced GSLBConfig below. Should that be the first object here? It feels awkward to reference an object that has not been introduced yet in this section.

File: docs/AMKO/crds/gdp.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    - cluster: cluster1-admin    // cluster names are kubernetes cluster contexts
+    - cluster: cluster2-admin
+ 
+  trafficSplit:
+    - cluster: cluster1
+      weight: 8
+    - cluster: cluster2
+      weight: 2
+```
+1. `namespace`: an important piece here, as a GDP object created in `avi-system` namespace is recognised and all other GDP objects created in other namespaces are ignored.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gdp.md:29]: Rules:
1. Namespace of this object must be `avi-system`

File: docs/AMKO/crds/gdp.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+2. `matchRules`: List of selection policy rules. If a user wants to select certain objects in a namespace (mentioned in `namespace`), they have to add those rules here. A typical `matchRule` looks like:
```
Comment (sudswasavi) [inline docs/AMKO/crds/gdp.md:30]: `matchRules`: This allows users to select objects using either application labels (configured as labels on Ingress/Route objects) or via namespace labels (configured as labels on the namespace objects). If both labels are provided, then AMKO will look for app labels inside the namespaces selected by the namespaceSelector.

File: docs/AMKO/crds/gdp.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+      label:
+        app: gslb
+    namespaceSelector:
+      label:
+        ns: prod
+```
+
+3. `matchClusters`: List of clusters on which the above `matchRules` will be applied on. The member object of this list are cluster contexts of the individual k8s/openshift clusters.
+
+4. `trafficSplit` is required if we want to route a certain percentage of traffic to certain objects in a certain cluster. These are weights and the range for them is 1 to 20.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gdp.md:77]: How many certains? :)

File: docs/AMKO/crds/gdp.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+### Notes
+* A GDP object must be created in the `avi-system` namespace. GDP objects in all ther namespaces will *not* be considered. For now, AMKO supports only one GDP object in the entire cluster. Any other additonal GDP objects will be ignored.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gdp.md:80]: s/ther/other

Comment (sudswasavi) [reply docs/AMKO/crds/gdp.md:80]: I think we have stated this above that GDP has to be created in avi-system.

Comment (sudswasavi) [reply docs/AMKO/crds/gdp.md:80]: You can mention the limitation as Limitation instead of Notes.

File: docs/AMKO/crds/gdp.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+* A GDP object is created as part of `helm install`. User can then edit this GDP object to modify their selection of objects.
+* Deletion of a GDP rule will trigger all the objects to be again checked against the remaining set of rules.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gdp.md:82]: This entire section is not needed IMHO.

File: docs/AMKO/crds/gslbconfig.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+7. `gslbLeader.controllerIP`: The GSLB leader IP address or the hostname along with the port number, if any.
+8. `memberClusters`: The kubernetes/openshift cluster contexts which are part of this GSLB cluster. See [here](#Multi-cluster kubeconfig) to create contexts for multiple kubernetes clusters.
+9. `globalServiceNameSource`: The basis on which a GSLB service is created and named. For now, the supported type is `HOSTNAME`. This means that the any ingress which shares a host name across clusters will be placed on the same GSLB service.
+10. `domainNames`: Supported GSLB subdomains.
+11. `refreshInterval`: This is an internal cache refresh time interval, on which syncs up with the AVI objects and checks if a sync is required.
+12. `logLevel`: Define the log level that the amko pod prints. The allowed levels are: `[INFO, DEBUG, WARN, ERROR]`.
+
+### Notes
+* Only one GSLBConfig object is allowed.
+* If using `helm install`, the GSLB config object will be created for you, just provide the right parameters in `values.yml`.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gslbconfig.md:39]: Above sentence edit:
A gslbconfig object is created by picking up values from the values.yaml

File: docs/AMKO/crds/gslbconfig.md [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+### Editability
+The only field that's editable today is the `logLevel` field. What this means is, a user can just change the logLevel in the runtime and the changes will take effect.
```
Comment (sudswasavi) [inline docs/AMKO/crds/gslbconfig.md:42]: We don't need this section again.

**Comment by hemantavi** [comment]
@sudswasavi took care of the review comments.

**Pull request merged**